### DISCUSSION TITLE
Unable to close Console Login Window after Successful login

### DISCUSSION THREAD

**Discussion by GitHubDroid** [tui-cs/Terminal.Gui#2782] [Q&A]
First let me say this project is impressive and I would like to try and develop on top of it.  

I am trying utilize the Console Login.   What I am trying to do is display a windows form after login.
The following seems to work for opening a Windows Form after login, but the Console Login Window stays open.

 // When login button is clicked display a message popup then Windows Form
            btnLogin.Clicked += () => {
                if (usernameText.Text == "admin" && passwordText.Text == "password")
                {
                    MessageBox.Query("Logging In", "Login Successful", "Ok");
                    Application.RequestStop();

                    System.Windows.Forms.Application.Run(new Form1());

![1](https://github.com/gui-cs/Terminal.Gui/assets/4502975/e90e5535-cdbf-4b8e-9bc2-57fc1366e944)

Any help or insight would be greatly appreciated.

**Comment by BDisp** [comment]
You must run it asynchronously. Do the cursor go to the next line after you call `System.Windows.Forms.Application.Run(new Form1());`? If you can provide some more code will be very helpful. Also don't forget to call `Application.Shutdown()` at the end.

**Reply by GitHubDroid** [reply]
The following is what I have in my Program.cs and LoginWindow.cs:

-- Program.cs


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Terminal.Gui;
using System.Windows.Forms;

namespace COY_Console
{
    class Program : System.Windows.Forms.Form
    {
        static void Main(string[] args)
        {
            Terminal.Gui.Application.Init();
            //Terminal.Gui.Application.Run<LoginWindow>();

            Terminal.Gui.Application.Run(new LoginWindow());
            //System.Console.WriteLine($"Username: {((LoginWindow)Terminal.Gui.Application.Top).usernameText.Text}");

            // Before the application exits, reset Terminal.Gui for clean shutdown
            Terminal.Gui.Application.Shutdown();
        }
    }
}

--LoginWindow.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Principal;
using System.Text;
using System.Threading.Tasks;
using COY_Console;
using Label = Terminal.Gui.Label;
using Button = Terminal.Gui.Button;
using MessageBox = Terminal.Gui.MessageBox;
using Application = Terminal.Gui.Application;
using Window = Terminal.Gui.Window;
using TextField = Terminal.Gui.TextField;
using Pos = Terminal.Gui.Pos;
using Dim = Terminal.Gui.Dim;
using Forms = System.Windows.Forms;
using Terminal.Gui;

namespace COY_Console
{
    public class LoginWindow : Window
    {
        
        public TextField usernameText;
        
        public LoginWindow()
        {
            Title = "Login to COY Console (Ctrl+Q to quit)";

            // Create input components and labels
            var usernameLabel = new Label()
            {
                Text = "Username:"
            };

            usernameText = new TextField("")
            {
                // Position text field adjacent to the label
                X = Pos.Right(usernameLabel) + 1,

                // Fill remaining horizontal space
                Width = Dim.Fill(),
            };

            var passwordLabel = new Label()
            {
                Text = "Password:",
                X = Pos.Left(usernameLabel),
                Y = Pos.Bottom(usernameLabel) + 1
            };

            var passwordText = new TextField("")
            {
                Secret = true,
                // align with the text box above
                X = Pos.Left(usernameText),
                Y = Pos.Top(passwordLabel),
                Width = Dim.Fill(),
            };

            // Create login button
            var btnLogin = new Button()
            {
                Text = "Login",
                Y = Pos.Bottom(passwordLabel) + 1,
                // center the login button horizontally
                X = Pos.Center(),
                IsDefault = true,
            };

            // When login button is clicked display a message popup then Windows Form
            btnLogin.Clicked += () => {
                if (usernameText.Text == "admin" && passwordText.Text == "password")
                {
                    MessageBox.Query("Logging In", "Login Successful", "Ok");
                    Application.RequestStop();

                    System.Windows.Forms.Application.Run(new Form1());

                   
                    
                    //Application.Run<Wizard>();

                   //Application.RequestStop();

                    


                }
                else
                {
                    MessageBox.ErrorQuery("Logging In", "Incorrect username or password", "Ok");
                }

                //Application.Top.Clear();
            };

            // Add the views to the Window
            Add(usernameLabel, usernameText, passwordLabel, passwordText, btnLogin);
        }

    }

    }

    
 Does this help?

**Comment by tznind** [comment]
Try adding Application.Shutdown after Application.RequestStop

**Reply by GitHubDroid** [reply]
That was actually the first thing I tried.   No success.

**Comment by tznind** [comment]
Winforms and terminal don't play well in same process.  I think launching
the winforms window will block the console thread.

You could start the winforms as a separate process with Process.Start?

**Comment by GitHubDroid** [comment]
The following seemed to close the LoginWindow. 

I added the following method:

          void frmNewFormThread()
            {
                System.Windows.Forms.Application.Run(frmNewForm);
            }

 The newThread.Start calls the above method below:


            var frmNewForm = new Form1();
            var newThread = new System.Threading.Thread(frmNewFormThread);

            // When login button is clicked display a message popup then Windows Form
            btnLogin.Clicked += () => {
                if (usernameText.Text == "admin" && passwordText.Text == "password")
                {
                    MessageBox.Query("Logging In", "Login Successful", "Ok");
                    Application.RequestStop();

                    
                    newThread.SetApartmentState(System.Threading.ApartmentState.STA);
                    newThread.Start();

                    //System.Windows.Forms.Application.Run(new Form1());

                }
                else
                {
                    MessageBox.ErrorQuery("Logging In", "Incorrect username or password", "Ok");
                }

                Application.Top.Clear();
            };

Does this seem like a feasible way to do it ?  



So my LoginWindow.cs looks like this:

    public class LoginWindow : Window
    {
        

        public TextField usernameText;
        

        

        public LoginWindow()
        {

            
            Title = "Login to COY Console (Ctrl+Q to quit)";

            // Create input components and labels
            var usernameLabel = new Label()
            {
                Text = "Username:"
            };

            usernameText = new TextField("")
            {
                // Position text field adjacent to the label
                X = Pos.Right(usernameLabel) + 1,

                // Fill remaining horizontal space
                Width = Dim.Fill(),
            };

            var passwordLabel = new Label()
            {
                Text = "Password:",
                X = Pos.Left(usernameLabel),
                Y = Pos.Bottom(usernameLabel) + 1
            };

            var passwordText = new TextField("")
            {
                Secret = true,
                // align with the text box above
                X = Pos.Left(usernameText),
                Y = Pos.Top(passwordLabel),
                Width = Dim.Fill(),
            };

            // Create login button
            var btnLogin = new Button()
            {
                Text = "Login",
                Y = Pos.Bottom(passwordLabel) + 1,
                // center the login button horizontally
                X = Pos.Center(),
                IsDefault = true,
            };

            var frmNewForm = new Form1();
            var newThread = new System.Threading.Thread(frmNewFormThread);
            // When login button is clicked display a message popup then Windows Form
            btnLogin.Clicked += () => {
                if (usernameText.Text == "admin" && passwordText.Text == "password")
                {
                    MessageBox.Query("Logging In", "Login Successful", "Ok");
                    Application.RequestStop();

                    
                    newThread.SetApartmentState(System.Threading.ApartmentState.STA);
                    newThread.Start();

                    //System.Windows.Forms.Application.Run(new Form1());

                }
                else
                {
                    MessageBox.ErrorQuery("Logging In", "Incorrect username or password", "Ok");
                }

                Application.Top.Clear();
            };

           

            // Add the views to the Window
            Add(usernameLabel, usernameText, passwordLabel, passwordText, btnLogin);


            //method to open winform in seperate process/thread
            void frmNewFormThread()
            {

                System.Windows.Forms.Application.Run(frmNewForm);

            }

        }

       

    }

**Reply by BDisp** [reply]
How do you referenced `System.Windows.Forms` on a `Console App`?

Edit:
If you have it on a github repo is better.

**Comment by GitHubDroid** [comment]
I will try and start a repo.  In the meantime I will post the files soon.  Stand by.  Thanks again for any assistance and insight.

**Reply by BDisp** [reply]
This repo https://github.com/BDisp/TerminalGuiOpenWinForm works on Windows and Linux. Take a look, please.

**Reply by GitHubDroid** [reply]
Link got a 404

**Reply by BDisp** [reply]
Sorry by default VS2022 made it private. Try now please.

**Reply by BDisp** [reply]
@GitHubDroid did my repo sample satisfy your needs?
